federation/executor.go

```diff
@@ -0,0 +1,323 @@
+// The planner allows it to coordinate the subqueries being sent to the federated servers
+type Executor struct {
+	Executors map[string]ExecutorClient
+	planner   *Planner  
+}
+
+func fetchSchema(ctx context.Context, e ExecutorClient) ([]byte, error) {
+	query, err := graphql.Parse(introspection.IntrospectionQuery, map[string]interface{}{})
+	if err != nil {
+		return nil, err
+	}
+
+	return e.Execute(ctx, query)
+}
+
+
+func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Executor, error) {
+	// Fetches the schemas from the executors clients
+	schemas := make(map[string]*introspectionQueryResult)
+	for server, client := range executors {
+		schema, err := fetchSchema(ctx, client)
+		if err != nil {
+			return nil, fmt.Errorf("fetching schema %s: %v", server, err)
+		}
+
+		var iq introspectionQueryResult
+		if err := json.Unmarshal(schema, &iq); err != nil {
+			return nil, fmt.Errorf("unmarshaling schema %s: %v", server, err)
+		}
+
+		schemas[server] = &iq
+	}
+
+	types, err := convertSchema(schemas)
+	if err != nil {
+		return nil, fmt.Errorf("converting schema error: %v", err)
+	}
+
+	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
+	introspectionServer := &Server{schema: introspectionSchema}
+
+	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
+	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionServer.schema))
+
+	var iq introspectionQueryResult
+	if err := json.Unmarshal(schema, &iq); err != nil {
+		return nil, fmt.Errorf("unmarshaling introspection schema: %v", err)
+	}
+
+	schemas["introspection"] = &iq
+	types, err = convertSchema(schemas)
+	if err != nil {
+		return nil, err
+	}
+
+	flattener, err := newFlattener(types.Schema)
+	if err != nil {
+		return nil, err
+	}
+
+	// The planner is aware of the merged schema and what executors 
+	// know about what fields
+	planner := &Planner{
+		schema:    types,
+		flattener: flattener,
+	}
+
+	return &Executor{
+		Executors: executors,
+		planner:   planner,
+	}, nil
+
+}
+
+// Runs a subquery on a specified service and returns the results
+func (e *Executor) runOnService(ctx context.Context, service string, typName string, keys []interface{}, kind string, selectionSet *graphql.SelectionSet) ([]interface{}, error) {
+	schema := e.Executors[service]
+
+	isRoot := keys == nil
+	if !isRoot {
+		selectionSet = &graphql.SelectionSet{
+			Selections: []*graphql.Selection{
+				{
+					Name:  "__federation",
+					Alias: "__federation",
+					Args:  map[string]interface{}{},
+					SelectionSet: &graphql.SelectionSet{
+						Selections: []*graphql.Selection{
+							{
+								Name:  typName,
+								Alias: typName,
+								Args: map[string]interface{}{
+									"keys": keys,
+								},
+								SelectionSet: selectionSet,
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	// TODO: make sure that if this hangs we're still good?
```

Comment:
I'd imagine the caller could set a timeout on `ctx` so this code doesn't wait indefinitely on the RPC?